resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -640,34 +626,265 @@ protected boolean doIsResolved(InetSocketAddress address) {
 
     @Override
     protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocketAddress> promise) throws Exception {
-        byte[] bytes = NetUtil.createByteArrayFromIpAddressString(unresolvedAddress.getHostName());
-        if (bytes == null) {
-            final String hostname = IDN.toASCII(hostname(unresolvedAddress));
-            final int port = unresolvedAddress.getPort();
+        final byte[] bytes = NetUtil.createByteArrayFromIpAddressString(unresolvedAddress.getHostName());
+        if (bytes != null) {
+            // The unresolvedAddress was created via a String that contains an ipaddress.
+            promise.setSuccess(new InetSocketAddress(InetAddress.getByAddress(bytes), unresolvedAddress.getPort()));
+            return;
+        }
+
+        final String hostname = hostname(unresolvedAddress);
+        final int port = unresolvedAddress.getPort();
+
+        if (!doResolveCached(hostname, port, promise)) {
+            doResolveUncached(hostname, port, promise);
+        }
+    }
+
+    private boolean doResolveCached(String hostname, int port, Promise<InetSocketAddress> promise) {
+        final List<DnsCacheEntry> cachedEntries = resolveCache.get(hostname);
+        if (cachedEntries == null) {
+            return false;
+        }
 
-            final DnsNameResolverContext ctx = new DnsNameResolverContext(this, hostname, port, promise);
+        InetAddress address = null;
+        Throwable cause = null;
+        synchronized (cachedEntries) {
+            final int numEntries = cachedEntries.size();
+            assert numEntries > 0;
 
-            ctx.resolve();
+            if (cachedEntries.get(0).cause() != null) {
+                cause = cachedEntries.get(0).cause();
+            } else {
+                // Find the first entry with the preferred address type.
+                for (InternetProtocolFamily f : resolveAddressTypes) {
+                    for (int i = 0; i < numEntries; i++) {
+                        final DnsCacheEntry e = cachedEntries.get(i);
+                        if (f.addressType().isInstance(e.address())) {
+                            address = e.address();
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        if (address != null) {
+            promise.trySuccess(new InetSocketAddress(address, port));
```

Comment:
Should we somehow respect the return values of trySuccess(...) and tryFailure(...) ?